client/api/src/backend.rs

```diff
@@ -536,3 +536,122 @@ pub fn changes_tries_state_at_block<'a, Block: BlockT>(
+/// component and Component), depending on future usage of this shared info.
+#[derive(Clone)]
+pub struct SharedPruningRequirements<Block: BlockT> {
+	shared: Arc<RwLock<(Vec<ComponentPruningRequirements<Block>>, usize)>>,
+	component: usize,
+}
+
+impl<Block: BlockT> Default for SharedPruningRequirementsSource<Block> {
+	fn default() -> Self {
+		SharedPruningRequirementsSource {
+			shared: Arc::new(RwLock::new((Vec::new(), 0))),
+		}
+	}
+}
+
+fn shared_finalized_headers_needed<Block: BlockT>(
+	shared: &Arc<RwLock<(Vec<ComponentPruningRequirements<Block>>, usize)>>,
+) -> PruningLimit<NumberFor<Block>> {
+	let mut result = PruningLimit::None;
+	for req in shared.read().0.iter() {
+		match &req.requires_finalized_header_up_to {
+			PruningLimit::Locked => return PruningLimit::Locked,
+			PruningLimit::None => (),
+			PruningLimit::Some(n) => {
+				if let PruningLimit::Some(p) = result {
+					if &p < n {
+						result = PruningLimit::Some(p);
+						continue;
+					}
+				}
+				result = PruningLimit::Some(n.clone());
+			},
+		}
+	}
+	result
+}
+
+impl<Block: BlockT> SharedPruningRequirementsSource<Block> {
+	/// Add a following requirement to apply.
+	/// Returns the shared requirement to use from this component.
+	pub fn next_instance(
+		&self,
+	) -> SharedPruningRequirements<Block> {
+		let req = ComponentPruningRequirements::default();
+		let index = {
+			let mut shared = self.shared.write();
+			let index = shared.0.len();
+			shared.0.push(req);
+			shared.1 += 1;
+			index
+		};
+		SharedPruningRequirements {
+			shared: self.shared.clone(),
+			component: index,
+		}
+	}
+
+	/// Resolve a finalized block headers to keep.
+	pub fn finalized_headers_needed(&self) -> PruningLimit<NumberFor<Block>> {
+		shared_finalized_headers_needed(&self.shared)
+	}
+}
+
+impl<Block: BlockT> SharedPruningRequirements<Block> {
+	/// Resolve a finalized block headers to keep.
+	pub fn finalized_headers_needed(&self) -> PruningLimit<NumberFor<Block>> {
+		shared_finalized_headers_needed(&self.shared)
+	}
+
+	/// Set new requirement on finalized headers.
+	/// Returns false if we do not have a handle on the shared requirement.
+	pub fn set_finalized_headers_needed(&self, limit: PruningLimit<NumberFor<Block>>) {
+		let index = self.component;
+		let mut shared = self.shared.write();
+		shared.0[index].requires_finalized_header_up_to = limit;
+	}
+}
+
+/// Individual pruning requirement for any substrate component.
+struct ComponentPruningRequirements<Block: BlockT> {
+	requires_finalized_header_up_to: PruningLimit<NumberFor<Block>>,
+}
+
+impl<Block: BlockT> Default for ComponentPruningRequirements<Block> {
+	fn default() -> Self {
+		ComponentPruningRequirements {
+			requires_finalized_header_up_to: PruningLimit::None,
+		}
+	}
+}
+
+/// Define a block number limit to apply.
+#[derive(Eq, PartialEq)]
+pub enum PruningLimit<N> {
+	/// Ignore.
+	None,
+	/// The component require at least this number
```

Comment:
Iiuc from the code, `N` here is actually number of oldest block that shouldn't be pruned. If I'm right, then the comment looks wrong - i.e. some component may need last 1024 headers, but we may be at block 1_000_000 => N would actually be `1_000_000 - 1024`